### ISSUE TITLE
Random Crashes on Linux with Torch/GPU

### ISSUE THREAD

**Issue by VisualMelon** [DiffSharp/DiffSharp#183]
I'm getting random crashes with Torch/GPU under x64 RHEL6 Linux, GTX 1060 with Cuda 11. The stack trace is always the same, but it only occasionally occurs (same code, same seed). The error clearly tells me to report to PyTorch, but I don't understand what is happening, so I won't do that just yet.

I've not dived into Torch/LibTorch yet (it's taken me long enough to slim down the not-so-minimal sometimes-reproducing example), but I shall do when I have time. Debugging on the linux instance is essentially impossible, and the crashes occur thousands of iterations apart, so I'm not hopeful. I don't have another linux box on which I can do proper debugging.

The same (pre significant trimming) code has been running reliably on my Windows development machine with Torch/GPU, and seems to run fine on the same linux instance with Torch/CPU. I'm confident some of what is less is unnecessary... but I've not been able to remove anything and observe it failing, so I'm not certain (it could be making the crashes less frequent, or stop altogether).

## Stack Trace (with user-specific stuff removed)

    Unhandled exception. System.Runtime.InteropServices.ExternalException (0x80004005): s.isIntegral(false) INTERNAL ASSERT FAILED at /pytorch/aten/src/ATen/ScalarOps.h:22, please report a bug to PyTorch.  (scalar_to_tensor at /pytorch/aten/src/ATen/ScalarOps.h:22)
    frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x46 (0x2b68bb5f6536 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libc10.so)
    frame #1: <unknown function> + 0xb0ec41 (0x2c2ae0b0ec41 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #2: <unknown function> + 0xb64a64 (0x2c2ae0b64a64 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #3: at::native::add(at::Tensor const&, c10::Scalar, c10::Scalar) + 0x28 (0x2c2ae0b64c38 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #4: <unknown function> + 0x1135772 (0x2c2ae1135772 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #5: <unknown function> + 0x10c5b32 (0x2c2ae10c5b32 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #6: <unknown function> + 0x2c0b8b7 (0x2c2ae2c0b8b7 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #7: <unknown function> + 0x10c5b32 (0x2c2ae10c5b32 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #8: at::Tensor c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&, c10::Scalar, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar, c10::Scalar) const + 0x15c (0x2b68bb15e11c in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #9: at::Tensor c10::Dispatcher::callUnboxed<at::Tensor, at::Tensor const&, c10::Scalar, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar, c10::Scalar) const + 0x150 (0x2b68bb15df80 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #10: THSTensor_addS + 0x109 (0x2b68bb12f929 in <redacted>/bin/Debug/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #11: [0x2b683ee5e358]

    at TorchSharp.Torch.CheckForErrors()
    at DiffSharp.Backends.Torch.TorchRawTensor.get_DeviceType()
    at DiffSharp.Backends.Torch.TorchRawTensor.get_Device()
    at DiffSharp.Backends.Torch.TorchRawTensor.MakeLike(TorchTensor tt, FSharpOption`1 shape, FSharpOption`1 dtype, FSharpOption`1 device)
    at DiffSharp.Backends.Torch.TorchRawTensor.AddTT0(RawTensor t2)
    at DiffSharp.Tensor.op_Addition(Tensor a, Tensor b)
    at <StartupCode$DiffSharp-Core>.$Tensor.push@1936(FSharpList`1 ts)
    at DiffSharp.Tensor.reversePush(Tensor value)
    at DiffSharp.Tensor.reverse(FSharpOption`1 value, FSharpOption`1 zeroDerivatives)
    at Program.main(String[] argv) in <redacted>/Program.fs:line 32

(The line in Program.fs is `combinedLoss.reverse()`)

### Sadness.fsproj

    <Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>netcoreapp3.1</TargetFramework>
    </PropertyGroup>
    <ItemGroup>
        <Compile Include="Program.fs" />
    </ItemGroup>
    <ItemGroup>
        <PackageReference Include="DiffSharp-cuda-windows" Version="1.0.0-preview-247523274" />
        <PackageReference Include="DiffSharp-cuda-linux" Version="1.0.0-preview-247523274" />
    </ItemGroup>
    </Project>

### Program.fs

    open DiffSharp

    [<EntryPoint>]
    let main argv =
        dsharp.config(backend=Backend.Torch, device=Device.GPU)
        dsharp.seed(1)

        // model
        let m = (dsharp.randn [12; 16]).reverseDiff()
        let s = dsharp.randn [50; 16]
        let l1 = dsharp.tensor 1e-3
        let l2 = dsharp.tensor 0

        for i in 1..70 do
            printfn "Outer %i" i
            for j in 1..2000 do
                let sr = s.matmul(m.transpose()).matmul(m)
                let loss = ((s - sr) * (s - sr)).sum()
                let mag = m.abs();
                let regLoss = (mag.sum() * l1) + ((mag * mag).sum() * l2)
                let combinedLoss = loss + regLoss
                combinedLoss.reverse()

        printfn "End"

        0

### Notes

It's all linear operations, so there are only so many operations in `reversePush` that might apply. Because it's a `TT0` operation, I would have expected it to occur with just the `regLoss` (i.e. just `regLoss.reverse()`, because the only constants are `l1` and `l2`), but I haven't observed that.

All tests pass on the same instance with the exception of the MNIST tests because it doesn't have internet access and I'm too lazy to move the data over.

**Comment by VisualMelon** [comment]
If I reference a local copy of the current DiffSharp dev branch and run as debug (so a debug build of DiffSharp), then the problems seem to stop.

If I run the same thing as configuration the error returns (and I had some memory usage problems, but that's hopefully unrelated). This also gives a somewhat more informative stack trace:

    Unhandled exception. System.Runtime.InteropServices.ExternalException (0x80004005): s.isIntegral(false) INTERNAL ASSERT FAILED at /pytorch/aten/src/ATen/ScalarOps.h:22, please report a bug to PyTorch.  (scalar_to_tensor at /pytorch/aten/src/ATen/ScalarOps.h:22)
    frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x46 (0x2b71e4092536 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libc10.so)
    frame #1: <unknown function> + 0xb0ec41 (0x2b71f4b0ec41 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #2: <unknown function> + 0xb64a64 (0x2b71f4b64a64 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #3: at::native::add(at::Tensor const&, c10::Scalar, c10::Scalar) + 0x28 (0x2b71f4b64c38 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #4: <unknown function> + 0x1135772 (0x2b71f5135772 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #5: <unknown function> + 0x10c5b32 (0x2b71f50c5b32 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #6: <unknown function> + 0x2c0b8b7 (0x2b71f6c0b8b7 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #7: <unknown function> + 0x10c5b32 (0x2b71f50c5b32 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libtorch_cpu.so)
    frame #8: at::Tensor c10::KernelFunction::callUnboxed<at::Tensor, at::Tensor const&, c10::Scalar, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar, c10::Scalar) const + 0x15c (0x2aafcfb7511c in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #9: at::Tensor c10::Dispatcher::callUnboxed<at::Tensor, at::Tensor const&, c10::Scalar, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar, c10::Scalar) const + 0x150 (0x2aafcfb74f80 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #10: THSTensor_addS + 0x109 (0x2aafcfb46929 in <redacted>/bin/Release/netcoreapp3.1/runtimes/linux-x64/native/libLibTorchSharp.so)
    frame #11: [0x2aaf537a0578]

    at TorchSharp.Torch.CheckForErrors()
    at DiffSharp.Backends.Torch.TorchRawTensor.get_DeviceType() in <redacted>/DiffSharp/src/DiffSharp.Backends.Torch/Torch.RawTensor.fs:line 89
    at DiffSharp.Backends.Torch.TorchRawTensor.get_Device()
    at DiffSharp.Backends.Torch.TorchRawTensor.MakeLike(TorchTensor tt, FSharpOption`1 shape, FSharpOption`1 dtype, FSharpOption`1 device) in <redacted>/DiffSharp/src/DiffSharp.Backends.Torch/Torch.RawTensor.fs:line 94
    at DiffSharp.Backends.Torch.TorchRawTensor.AddTT0(RawTensor t2) in <redacted>/DiffSharp/src/DiffSharp.Backends.Torch/Torch.RawTensor.fs:line 432
    at DiffSharp.Tensor.op_Addition(Tensor a, Tensor b) in <redacted>/DiffSharp/src/DiffSharp.Core/Tensor.fs:line 583
    at <StartupCode$DiffSharp-Core>.$Tensor.push@1936(FSharpList`1 ts)
    at DiffSharp.Tensor.reversePush(Tensor value) in <redacted>/DiffSharp/src/DiffSharp.Core/Tensor.fs:line 2097
    at DiffSharp.Tensor.reverse(FSharpOption`1 value, FSharpOption`1 zeroDerivatives) in <redacted>/DiffSharp/src/DiffSharp.Core/Tensor.fs:line 1814

I probably won't need GPU in the near future, so this isn't a massive concern for me at the moment, but I'll keep looking.

**Comment by dsyme** [comment]
Thanks, that's a good starting repro

I guess next steps might be to 

1. reduce to a loop of TorchRawTensor operations not involving gradients 

2. Then reduce to a TorchSharp test case not involving DiffSharp at all

3. Maybe finally to a C++ test case

This would determine at which layer the problem is and the relevant code to review 

Also we'll upgrade to Libtirch 1.6.0 so beyond soon, possible that might fix this sort of thing. Would be worth trawling the pytorch bugs to see if there's anything there

**Comment by gbaydin** [comment]
I had a quick look. 

The error seems to be happening when reading the DeviceType of a Torch rawtensor:

https://github.com/DiffSharp/DiffSharp/blob/d2841215865902cd3cc2100e7c7409d71ad6a1b6/src/DiffSharp.Backends.Torch/Torch.RawTensor.fs#L89

Through TorchSharp:

https://github.com/xamarin/TorchSharp/blob/f456b0d1c162b6b7b4e7947df61fff3fea984d7d/src/TorchSharp/Tensor/TorchTensor.cs#L336

this ultimately ends up here:

https://github.com/pytorch/pytorch/blob/da7863f46bff365499730244a3198a9d4b201f7c/aten/src/ATen/ScalarOps.h#L22

and the `isIntegral` check fails. Which makes me wonder whether this is an int32/int64 issue. I'm not sure at this point.

**Comment by VisualMelon** [comment]
Sorry I haven't provided any updates: the compute provision is down at the moment, and will be busy with other work once it is up, so it may be a while until I can perform any proper tests again. Once I've verified whether the problem exists with the newer version of TorchSharp (not sure what's changed) I'll go back to trying to build a (minimal) failing example from the ground up, rather than trying to remove things from my existing code.

**Comment by kevmal** [comment]
@VisualMelon Maybe confirm CUDA is fully up to date. I noticed there's a recent (Aug 26) patch to CUDA 10.2:
> This patch fixes an issue in the cuBLAS library bundled in CUDA 10.2 which caused silent corruption of data in uncommon edge cases

And the `matmul` op is most likely using cuBLAS.

**Comment by VisualMelon** [comment]
@kevmal thanks for the tip, though, I'm using Cuda 11 (I probably should have mentioned that earlier). Could you provide a link to the documentation for that fix? I can't find it, or an analogue for Cuda 11. I'll check the exact version when the system is working again, though I won't be able to update it, at least no in a hurry.

**Comment by kevmal** [comment]
For 10.2:
https://docs.nvidia.com/cuda/archive/10.2/cuda-toolkit-release-notes/index.html#cuda-libraries-resolved-issues

For 11 there's an August update but I'm not sure if the resolved issues are relevant. I couldn't reproduce the error (CUDA 10.2/Ubuntu16/FSI .NET 5 preview) when I tried about a week ago. Ill try again with CUDA 11.0 when I get a chance, If I can hit the issue I can check if updating would make a difference.

**Comment by VisualMelon** [comment]
@kevmal thanks for the links, and for taking the time to look at this. I'm sorry I can't be more help off-hand.

**Comment by VisualMelon** [comment]
I've not made any progress on this, but I can confirm the random crashes still occur after the 1.7 update.

**Comment by dsyme** [comment]
> I've not made any progress on this, but I can confirm the random crashes still occur after the 1.7 update. (Same hardware as the OP)

Still CUDA 11?

**Comment by VisualMelon** [comment]
Yes, still CUDA 11.0.

**Comment by dsyme** [comment]
@VisualMelon Could you recheck this now that DiffSharp uses CUDA 11.0? Thanks

**Comment by VisualMelon** [comment]
Yeah, I'll try to get that done today; do prod me if I don't get back to you in the next couple of days.

**Comment by VisualMelon** [comment]
After some fiddling and confusion, I have my not-so-minimal example running on the Linux server, and it hasn't immediately crashed; if it survives a few thousand iterations I think we can close this (may be a while before I get the results from those, so I'll get back to you on that)

I couldn't build it with .NET 5; using .NET 6 Preview5 (is this expected? can run with .NET 5 on my windows box, not tried on the Linux server)

Notes:
 - Same Linux box as before, with CUDA 11.0 installed (as reported by nvcc; same on my Windows box)
 - Using the DiffSharp-cuda-linux package Version 1.0.0-preview-987646120
 - Ran all the tests on my Window box with GPU while I was waiting, and they all passed

**Comment by dsyme** [comment]
> I couldn't build it with .NET 5; using .NET 6 Preview5 (is this expected? can run with .NET 5 on my windows box, not tried on the Linux server)

Yes, .NET 6 Preview is required to successfully restore the Linux CUDA packages https://github.com/DiffSharp/DiffSharp/#available-packages-and-backends

**Comment by dsyme** [comment]
Thank you so much for checking all this!

**Comment by VisualMelon** [comment]
4 repeats all completed without issue, so I'm happy for this to be closed.

**Comment by dsyme** [comment]
Great!
